README_B2C.md

```diff
@@ -0,0 +1,167 @@
+languages:
+- python
+- html
+products:
+- azure-active-directory
+description: "This sample demonstrates a Python web application calling a Microsoft Graph that is secured using Azure Active Directory."
+urlFragment: ms-identity-python-webapp
+---
+# Integrating B2C feature of Microsoft Identity Platform with a Python web application
+
+## About this sample
+
+> This sample was initially developed as a web app to demonstrate how to
+> [Integrate Microsoft Identity Platform with a Python web application](https://github.com/Azure-Samples/ms-identity-python-webapp/blob/master/README.md).
+> The same code base can also be used to demonstrate how to
+> Integrate B2C feature of Microsoft Identity Platform with a Python web application.
+> All you need is some different steps to register your app in your own B2C tenant,
+> and then feed those different settings into the configuration file of this sample.
+
+This sample covers the following:
+
+* Update the application in Azure AD B2C
+* Configure the sample to use the application
+* Enable authentication in a web application using Azure Active Directory B2C
+* Access a web API using Azure Active Directory B2C
+
+
+### Overview
+
+This sample demonstrates a Python web application that signs-in users with the Microsoft identity platform and calls the Microsoft Graph.
+
+1. The python web application uses the Microsoft Authentication Library (MSAL) to obtain an access token from the Microsoft identity platform (formerly Azure AD v2.0):
+2. The access token is used as a bearer token to authenticate the user when calling the Microsoft Graph.
+
+![Overview](./ReadmeFiles/topology.png)
+
+
+## Prerequisite
+
+1. [Create an Azure Active Directory B2C tenant](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-create-tenant)
+1. [Register an application in Azure Active Directory B2C](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-register-applications).
+1. [Create user flows in Azure Active Directory B2C](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-create-user-flows)
+1. Have [Python 2.7+ or Python 3+](https://www.python.org/downloads/) installed
+
+
+## Update the application
+
+In the tutorial that you completed as part of the prerequisites, you added a web application in Azure AD B2C.
+To enable communication with the sample in this tutorial, you need to add a redirect URI to the application in Azure AD B2C.
+
+1. Sign in to the [Azure portal](https://portal.azure.com/).
+1. Make sure you're using the directory that contains your Azure AD B2C tenant by selecting the **Directory + subscription** filter in the top menu and choosing the directory that contains your tenant.
+1. Choose **All services** in the top-left corner of the Azure portal, and then search for and select **Azure AD B2C**.
+1. Select **Applications**, and then select the *webapp1* application.
+1. Under **Reply URL**, add something like `http://localhost:5000/getAToken`.
```

Comment:
>localhost:5000 [](start = 51, length = 14)

localhost is not supported w/b2c yet....or did you get it to work?

Reply:
@jennyf19 Yup, localhost is supported. We use this in the tutorials when we direct readers to configure the sample(s) to run locally. For example, in the [Update the application](https://docs.microsoft.com/azure/active-directory-b2c/active-directory-b2c-tutorials-web-app#update-the-application) section of the ASP.NET tutorial part 1.

Reply:
@rayluo I'm wondering if you can specify the required settings here without including the actual step-by-step in the README, and instead, include a link to the article on docs.microsoft.com?

With the rate at which portal UI changes, keeping these steps up to date is a never-ending game of whack-a-mole. And, in fact, they're already out of date now with the fact that the **All services** doesn't appear by default, anymore--you have to make the left-hand menu "sticky," or select the hamburger to expand it. (Another big update I have to go through all of our docs and fix in the next week).

Reply:
@jennyf19 Yes, `localhost` would work. In our shared internal test environment `fabrikamb2c`, we tested it with `http://localhost:8080/msal4jsample` would work. Proof:

![image](https://user-images.githubusercontent.com/821550/68043923-a6c79300-fc93-11e9-9cea-9413f3b62a9d.png)

In general, there is no real technical reason to disallow localhost. It is pretty much necessary when app developers are still developing locally. For what it's worth, sometimes the UI might want to force all `https://`, but then a localhost typically does not have a cert, so `https://localhost` may be accepted by portal but not actually work in local web server, hence caused some confusion.

---

@mmacy , those step-by-step were largely copied from [existing tutorial](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-tutorials-web-app). I totally agree with you, that we should NOT write our doc in this whack-a-mole way (LOL). Do you happen to know an existing, generic article on docs.microsoft.com that we can link to?

Reply:
thanks @rayluo , i think it's the wildcards that is not allowed w/localhost yet. cool to see it working.

Reply:
@rayluo As far as a target for this:

>  Do you happen to know an existing, generic article on docs.microsoft.com that we can link to?

How about linking directly to the section of that article that includes the app update update steps, and specifying only what they should enter for their reply URL:

`https://docs.microsoft.com/azure/active-directory-b2c/active-directory-b2c-tutorials-web-app#update-the-application`

The risk, of course, is that we can't do redirects on a section, only on an article. So, if that article moves or the H2 is modified, your link will bust. I guess it's down to how vigilant can you be in keeping the steps up to date? The articles on docs are (or should be) updated on a more regular cadence than READMEs, so it's your call. Another thing to note is that the new app registration experience will be posted in that article on 04 NOV, so you'd pick that change up for "free" if you linked to that section.

Reply:
@mmacy Thanks for the suggestion. I prefer to not [deep link into that existing document](https://docs.microsoft.com/en-us/azure/active-directory-b2c/active-directory-b2c-tutorials-web-app#update-the-application). It is less about the concern of the H2 would be modified. It is more about that section contains specific instruction that was only applicable for that .Net sample.

How about this [latest commit](https://github.com/Azure-Samples/ms-identity-python-webapp/pull/4/commits/fd208774c4b4d84133f89d6c661e549325e7d584)? It just roughly re-link to the generic prerequisite, and tell the reader to make some change. After all, if they've been successfully followed the prerequisites, they should already know where/how to modify their app's registration.

Reply:
@mmacy @rayluo Great points being discussed in this PR! I vote to link to that section in the docs article.